libs/ui/src/accessible_controllers/navigation.ts

```diff
@@ -0,0 +1,60 @@
+export enum PageNavigationButtonId {
+  NEXT = 'next',
+  PREVIOUS = 'previous',
+}
+
+const FOCUSABLE_ELEMENT_SELECTORS = [
+  'button:not([aria-hidden="true"]):not([disabled]):not([tabindex="-1"])',
+  '[role="button"]:not([aria-hidden="true"]):not([disabled]):not([tabindex="-1"])',
+].join(', ');
+
+const HIDDEN_TAB_ENABLED_ELEMENT_SELECTORS = [
+  '[aria-hidden="true"] button',
+  '[aria-hidden="true"] [role="button"]',
+].join(', ');
+
+function getHiddenTabEnabledElements() {
+  return new Set(
+    document.querySelectorAll(HIDDEN_TAB_ENABLED_ELEMENT_SELECTORS)
+  );
+}
+
+/**
+ * Simulates browser `Tab`/`Shift+Tab` behavior by moving focus to the next
+ * focusable element in the document, in the specified {@link direction}.
+ */
+export function advanceElementFocus(direction: 1 | -1): void {
+  const hiddenTabEnabledElements = getHiddenTabEnabledElements();
+
+  const focusableElements = Array.from(
+    document.querySelectorAll<HTMLElement>(FOCUSABLE_ELEMENT_SELECTORS)
+  ).filter((e) => !hiddenTabEnabledElements.has(e));
```

Comment:
Are `hiddenTabEnabledElements` already filtered by the `not([aria-hidden="true"])` clauses in `FOCUSABLE_ELEMENT_SELECTORS`?

Reply:
This could probably use a more explicit naming - it's more along the lines of `tabEnabledElementsInHiddenBlocks`, which don't get caught by the `not([aria-hidden="true"])` and have to be queried via an ancestor query: `[aria-hidden="true"] button`

Will update naming and add clarifying comments to avoid confusion

Reply:
Thanks for clarifying!